visionarypm/__init__.py

```diff
@@ -106,35 +103,86 @@ def getConfig():
 
 
 def generate(master_password, keyword, cost, oLen=None, num_words=None):
+
     if not num_words:
         num_words = params['nwords']
-    dict_len = len(words)
-    entropy_per_word = math.log(dict_len, 2)
-    hash = codecs.encode(scrypt.hash(password=master_password, salt=keyword, N=1<<cost, buflen=64), 'hex').decode('utf-8')
-    available_entropy = len(hash) * 4
-    seed = int(hash, 16)
-    if (num_words * entropy_per_word) > available_entropy:
-        raise Exception("The output entropy of the specified hashfunc (%d) is too small." % available_entropy)
-    phrase = []
-    for i in range(num_words):
-        remainder = seed % dict_len
-        seed = seed // dict_len
-        phrase.append(words[int(remainder)].strip().decode('utf-8'))
-    readable_hash = " ".join(phrase).lower().capitalize()
+
+    def normal_password():
+        seed = scrypt.hash(password=master_password,
+                           salt=keyword,
+                           N=1<<cost,
+                           buflen=64)
+        return codecs.encode(seed, 'hex').decode('utf-8')
+
+    normal = normal_password()
+
+    # Courtesy of rubik
+    def complex_password():
+        # A random shuffle of string.punctuation + string.ascii_uppercase
+        to_add = '*>OI+D~}E{AF"`KXR\[J:-&#BQWN;$%TGPZ.=U,_HS()@L\'|<M]YCV/?!^'
+        required = (string.ascii_lowercase,
+                    string.ascii_uppercase,
+                    string.digits,
+                    string.punctuation)
+
+        def nth_perm(S, n):
+            '''Compute the n-th lexicographic permutation of the alphabet S.'''
+            p = []
+            for l in range(len(S) - 1, -1, -1):
+                i, n = divmod(n, math.factorial(l))
+                p.append(S[i])
+                S = S[:i] + S[i + 1:]
+            return p
+
+        res = []
+        perm_number = len(keyword) * sum(map(ord, keyword))
+        chosen_perm = nth_perm(to_add, perm_number % math.factorial(len(to_add)))
+        for i, c in enumerate(normal):
+            # 2/3 of the original is preserved
+            if i % 3:
+                res.append(c)
+                continue
+            # We index from the bottom of the string, because that's where the
+            # permutation changes most
+            res.append(chosen_perm[-((i + len(keyword)) % len(to_add) + 1)])
+        # There's a little possibility that the required chars aren't there
+        for i, req in enumerate(required):
+            if not set(req).intersection(res):
+                res[i] = req[0]
```

Comment:
Since `to_add` is shuffled, the probability of this happening is really low. But if it happens it chooses always the first character of the required set (`req[0]`). This is not very good, and the index instead of being zero should be calculated in some other way. Maybe from the original password itself?